junit5/src/main/java/cz/xtf/junit5/listeners/OpenshiftStateRecorder.java

```diff
@@ -0,0 +1,199 @@
+		try {
+			storeOpenshiftState(context.getDisplayName());
+		} catch(Exception newEx){
+			// if new exception is thrown, add it to original one so it is not lost
+			ex.addSuppressed(newEx);
+		}
+		throw ex;
+	}
+
+	// Log the openshift state if test fail
+	@Override
+	public void executionFinished(TestIdentifier testIdentifier, TestExecutionResult testExecutionResult) {
+		try {
+			if (FAILED.equals(testExecutionResult.getStatus()) && testIdentifier.isTest()) {
+				storeOpenshiftState(getTestDisplayName(testIdentifier));
+			}
+		} catch (IOException e) {
+			e.printStackTrace();
+		}
+	}
+
+	private void storeOpenshiftState(String testIdentifier) throws IOException {
+		final Path testLogDir = statusDir.resolve(testIdentifier);
+		Files.createDirectories(testLogDir);
+
+		storeNamespaceState(testLogDir, openShift);
+		storePods(testLogDir, openShift);
+
+		if (!namespace.equals(buildNamespace)){
+			storeNamespaceState(testLogDir, buildOpenShift);
+			storePods(testLogDir, buildOpenShift);
+		}
+	}
+
+	private void storePods(Path dir, OpenShift openshift){
+		for (Pod pod : openshift.getPods()) {
+			try {
+				openshift.storePodLog(pod, dir, pod.getMetadata().getName() + ".log");
+			} catch (IOException e) {
+				log.warn("IOException storing pod logs", e);
+			} catch (KubernetesClientException e) {
+				log.warn("KubernetesClientException getting pod logs", e);
+			}
+		}
+	}
+
+	private void storeNamespaceState(Path directory, OpenShift openShift) throws IOException {
+		String namespace = openShift.getNamespace();
+		BufferedWriter writer = new BufferedWriter(new FileWriter(directory.toAbsolutePath() + "/namespace-" + namespace + ".txt"));
+		writer.write("Namespace " + namespace + " state - " + LocalDateTime.now());
+		writer.newLine();
+
+		writer.write("Pods: (name ; status ; readyContainers/totalContainers ; containerRestart)");
+		writer.newLine();
+		for (Pod pod : openShift.getPods()) {
+			long readyContainers = pod.getStatus().getContainerStatuses().stream()
+						.filter(ContainerStatus::getReady).count();
+
+			writer.write(pod.getMetadata().getName()+ " ; " + pod.getStatus().getPhase()
+					+ " ; " + readyContainers + "/" + pod.getStatus().getContainerStatuses().size()
+					+ " ; " + pod.getStatus().getContainerStatuses().stream()
+						.map(ContainerStatus::getRestartCount)
+						.mapToInt(Integer::intValue)
+						.sum());
+			writer.newLine();
+		}
+		writer.newLine();
+
+
+		writer.write("DeploymentConfigs: (name - status)");
+		writer.newLine();
+		for (DeploymentConfig deploymentConfig : openShift.getDeploymentConfigs()) {
+			writer.write(deploymentConfig.getMetadata().getName()+ " replicas:" + deploymentConfig.getStatus().getReplicas() + " readyReplicas: " + deploymentConfig.getStatus().getReadyReplicas());
+			writer.newLine();
+		}
+		writer.newLine();
+
+
+		writer.write("Events:");
+		writer.newLine();
+		for (Event event : openShift.getEvents()){
```

Comment:
This will get all events in past N hours. N is cluster wide configurable, typically 2 hours.

Is it possible to filter just events which are connected to current test? E.g. by timestamp filtering?

If events will be not filtered just to current test, then it does not add much. There are already events saved once testsuite finished - by `EventsRecorder`.

Reply:
Nice catch I will look into this. Original purpose for this was to easily find out, when tests fails because of some image pull failure etc. 
Aren't events are also limited by number? I think I never got more that about 100, but I'm not sure about this.